### PR TITLE
Paper: Scikit-build-core: A modern build-backend for CPython C/C++/Fortran/Cython extensions

### PR DESCRIPTION
**Pull request by henryiii**
First draft. CC @jcfr and @thewtex.

_Editor_: Chris Calloway @cbcunc 

_Reviewers_: 
- Andy Terrel @aterrel
- Blaine Mooers @MooersLab

### CODE REVIEW

**Comment by ameyxd** [comment]
Hi @henryiii thanks for your submission! Can you update the DOI links to run checks?

File: papers/henry_schreiner/myst.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  title: Scikit-build-core
+  subtitle: A modern build-backend for CPython C/C++/Fortran/Cython extensions
+  # Authors should have affiliations, emails and ORCIDs if available
+  authors:
+    - name: Henry Schreiner
+      email: henryfs@princeton.edu
+      orcid: 0000-0002-7833-783X
+      affiliations:
+        - Princeton University
+      corresponding: true
```
Comment (jcfr) [inline papers/henry_schreiner/myst.yml:15]:
```suggestion
      corresponding: true
      roles:
        - writing
```

File: papers/henry_schreiner/myst.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    - name: Jean-Christophe Fillion-Robin
+      email: jcfr@kitware.com
+      orcid: 0000-0002-9688-8950
+      affiliations:
+        - Kitware Inc.
+      corresponding: true
```
Comment (jcfr) [inline papers/henry_schreiner/myst.yml:23]:
```suggestion
      corresponding: true
      roles:
        - writing
```

File: papers/henry_schreiner/myst.yml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    - name: Matt McCormick
+      email: matt.mccormick@kitware.com
+      orcid: 0000-0001-9475-3756
+      affiliations:
+        - Kitware Inc.
```
Comment (jcfr) [inline papers/henry_schreiner/myst.yml:30]:
```suggestion
        - Kitware Inc.
      roles:
        - writing
```

**Comment by jcfr** [comment]
Based on example mentioned in [this post](https://tex.stackexchange.com/questions/532049/how-to-cite-python-pep), the PEPs referenced in the checks have been cited[^1] using a `techreport` entry.

[^1]: https://github.com/henryiii/scipy_proceedings/blob/a52fd7bf80257f6201a622d14eeb7e08847f8df8/papers/henry_schreiner/mybib.bib#L43-L76

That said, I am not clear which DOI link to use in that particular case.

Is there way to add exceptions to allow the check to complete ?

_From the [instructions](https://github.com/scipy-conference/scipy_proceedings?tab=readme-ov-file#general-information-and-guidelines-for-authors)_:

> All citations that have DOIs should include those DOIs in the paper's references sectio

**Comment by ameyxd** [comment]
@jcfr, @henryiii you may be able to add citation keys you want to ignore if their DOIs don't exist in myst.yml under error_rules:

```error_rules:
  - rule: doi-exists
    severity: ignore
    keys:
      - abc
      - def01
 ```

File: papers/henry_schreiner/myst.yml [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    PEP: Python Enhancement Proposal
+    NSF: National Science Foundation
+  # A banner will be generated for you on publication, this is a placeholder
+  banner: banner.png
+  # The rest of the information shouldn't be modified
+  subject: Research Article
+  open_access: true
+  license: CC-BY-4.0
+  venue: Scipy 2024
+  date: 2024-07-10
```
Comment (fwkoch) [inline papers/henry_schreiner/myst.yml:49]: If these PEP citations do not have DOIs, this should be the required update:

```suggestion
  date: 2024-07-10
  error_rules:
    - rule: doi-exists
      severity: ignore
      keys:
        - pep517
        - pep518
        - pep621
        - pep660
```

**Comment by cbcunc** [comment]
As @jcfr is a co-author, I went ahead and committed his review suggestions in order to get the paper to pass all checks so that the paper is ready for review by the assigned reviewers. Thanks @jcfr and I hope to see you in Tacoma.

**Comment by jcfr** [comment]
Thanks @cbcunc for the help :pray:

And yes, also looking forward to see you in Tacoma. @thewtex will also be here.

**Comment by henryiii** [comment]
Thanks! @jcfr, I meant to give you permissions here, I can add them if you don’t already have them.

**Comment by henryiii** [comment]
Added both you and @thewtex, sorry I forgot!

**Comment by cbcunc** [comment]
Review reminders sent to @aterrel and @MooersLab

**Comment by aterrel** [comment]
Hello! I'll have some notes on the review this week, been on vacation and travel the last few.

**Comment by MooersLab** [comment]
# Review of *Scikit-build-core: A modern build-backend for CPython C/C++/Fortran/Cython extensions*

## General Comments
Hi!

I apologize to the authors for my tardy response.

I am new to Python extension building: I am not a domain expert.
Nonetheless, this paper is well-organized, precise, clear, and concise. 
After reading this paper, I want to use scikit-build-core to build a Python package,  
 

Do software names need to be in italics or bold?
Some package names are in different font, but the use of the different font does seem to me to be not consistent.

## Some very minor copy-editing changes to suggest.

## Introduction

I appreciate the respect that the authors show for the Oxford comma! Yah!
You can replace the first comma in the first sentence of paragraph two 
of the Introduction with an EM-dash (--) or colon to avoid confusion with the 
four other commas in the sentence.

In the second sentence, I suggest inserting "new system" between "this" and "provides"


## History of Packaging


- Paragraph 3, sentence one: *full featured* needs a hyphen.

- Paragraph 3, sentence one: comma before "even"

- Paragraph 3, sentence two: sdist and SDist are not consistent. 
Define here sadist with its first use rather than in the second paragraph of ### The simplest example.

- Paragraph 7, sentence two: Python only needs a hyphen.


## Scikit-build (classic)

last line: awkward array is written Awkward Array by the developers.

### Integration with external packages

Collapse this section into a single paragraph.
The first two sentences can be combined and rewritten as one sentence.

### Dual editable modes with automatic recompile.

Second to last sentence: replace *since* with *because*. 
Since is a weasel word when used to refer to causation. 
Just use *because* to be more direct. 


### The configuration system

First paragraph, last sentence: *JSONSchema* needs to be separated into two words.


Third paragraph, last sentence: replace *since* with *because*
Third paragraph, last sentence: "yet and" is confusing. Rewrite for clarity.


### Statistics
Two paragraph: There is a (TODO). Is it done?

**Comment by aterrel** [comment]
Independent Review Report
=========================

Reviewer: Andy Terrel

Department/Center/Division: Compute Products

Institution/University/Company: NVIDIA

Field of interest / expertise: Computer Science / Computational Math

Country: USA


Article reviewed: Scikit-build-core: A modern build-backend for CPython C/C++/Fortran/Cython extensions


GENERAL EVALUATION
------------------

Please rate the paper using the following criteria (please use the abbreviation
to the right of the description)::

  below     doesn't meet standards for academic publication
  meets     meets or exceeds the standards for academic publication
  n/a       not applicable

- Quality of the approach: meets

- Quality of the writing: meets

- Quality of the figures/tables: meets


SPECIFIC EVALUATION
-------------------

For the following questions, please respond with 'yes' or 'no'.  If you
answer 'no',  please provide a brief, one- to two-sentence explanation.

- Is the code made publicly available and does the article sufficiently
  describe how to access it?  

Yes

- Does the article present the problem in an appropriate context?
Yes

- Is the content of the paper accessible to a computational scientist
  with no specific knowledge in the given field?

Yes. The article does a good job of explaining the history and current state of building python software then gives a small example for reference.

- Does the paper describe a well-formulated scientific or technical
  achievement?

Yes, a coherent build backend that is used by numerous project is both a worthy goal and a technical achievement.

- Are the technical and scientific decisions well-motivated and
  clearly explained?

Yes

- Are the code examples (if any) sound, clear, and well-written?

Yes

- Is the paper factually correct?

Yes. Although a few phrases are not fact but supposition:

 "Python has a long history" I mean what does long mean not nearly as long as FORTRAN or the Euler equation.  "packaging isn’t something that was considered important" that may be the feeling of the community but what facts are supporting this statement. 
 "Leading to packages containing large numbers of unrelated modules (SciPy) to reduce the number of packages one had to figure out how to install" -> I don't think this was the motivation for SciPy or that the modules are "unrelated"

Additionally, if you are going to talk about wheel formats you should also talk about conda packages which predate the wheel and are widely used by the community. Especially since you include it in your figure.

If one is going to call out Rapids for "unusual requirements" and breaking PEP 621, it would be good to expound on those unusual requirements.

- Is the language and grammar of sufficient quality?

Yes

- Are the conclusions justified?

Yes, but claims that it is "easiest" and "first" on minor features (like cli builds and free-threading) are noise rather than the true engineering feat. 

- Is prior work properly and fully cited?

I think you need to include conda in your history section as it was a prior work solving a very similar problem.

- Should any part of the article be shortened or expanded? Please explain.

No.

- In your view, is the paper fit for publication in the conference proceedings?
  Please suggest specific improvements and indicate whether you think the
  article needs a significant rewrite (rather than a minor revision).
  
  Yes but please fix the (TODO)s

**Comment by henryiii** [comment]
Working on this now!

**Comment by henryiii** [comment]
@aterrel

> I don't think this was the motivation for SciPy 

"unrelated" is incorrect, I wanted to imply they didn't need to be together (not interdependent?). I think "distinct" is better. But otherwise I think this is exactly how Travis Oliphant describes SciPy ("His baby") in <https://www.youtube.com/watch?v=gFEE3w7F0ww&t=1372s>. It was the merging of several packages for ease of distribution. Ahh, that's also described in the SciPy paper, will move the reference up.

> if you are going to talk about wheel formats you should also talk about conda packages

Conda packages just run `pip install .` which itself builds a wheel then installs it. Conda is just a general purpose (not just Python) redistribution like Enthought that is just also modular. If you have to discuss conda when discussing wheels, why not every other package manager? Conda happens to have a lot of Python in its history, but it's a redistribution.

I do think it make sense to cover conda in the paragraph before, though, especially since it came first (which I knew but had somewhat forgotten).

> it would be good to expound on those unusual requirements

I thought that's exactly what I did in the following sentences? They need to change the name of the package based on the CUDA version it is compiled with and inject modified dependencies? I can expand but not sure what more to say. I did add a sentence about the discussions on this at PyCon 24's packaging summit.

>  but claims that it is "easiest" and "first" on minor features (like cli builds and free-threading)

The word `[Ee]siest` doesn't appear in the document. CLI builds - if you mean building without `build` or `pip`, but from some custom CLI tool, aren't supported yet, while maturin does support this, so we better not claim to be first there!

I removed the `(if not the first)` for free-threaded builds. But I wouldn't call free-threading a minor feature - it's the largest change to CPython ever. It's the first feature to force multiple builds since wide unicode, and it breaks the Stable ABI.

> conda in your history section as it was a prior work solving a very similar problem

It's there now, ~~but it doesn't solve a similar problem at all.~~ Conda can't build anything by itself; it uses your build system (including scikit-build-core) and distributes binary that don't require building. Ahh, what you mean is you can install via cmake commands instead of python ones, and if CMake puts things in the right spots with the right names, it can be a Conda-only distribution. Similar to how root on conda-forge works. But probably missing a `.dist-info` folder. Okay, fair, adding a paragraph.

**Comment by henryiii** [comment]
I think I addressed everything except the package names, should they be `package` or *package*?

**Comment by henryiii** [comment]
Ah, and sorry about the TODO. It takes several hours to run the statistics, and I started it, wrote the TODO, then completely forgot about it.

**Comment by aterrel** [comment]
@henryiii Thanks for addressing my comments. I think adding these small details to the great work the team has done will only strengthen the paper. I look forward to seeing the presentation.

**Review by aterrel — APPROVED** [review]

**Review by MooersLab — APPROVED** [review]
@bmooers approves this pull request. This is a very valuable contribution to the Proceedings.

**Comment by cbcunc** [comment]
I just want to address the statement in an earlier comment:

> Conda packages just run pip install .

Many conda packages just repackage a Pypi distribution. However, I am many other use grayskull to package straight to a conda package without Pypi ever being involved. Conda packages themselves don't just run a pip install, either, and are not redistributions.

**Comment by henryiii** [comment]
I realized this in the last paragraph of https://github.com/scipy-conference/scipy_proceedings/pull/939#issuecomment-2206785259.

Many conda packages are redistributions, though. Even if they don’t use pip, they can still be a redistribution.

**Pull request merged**